### PR TITLE
Add DashScope model wrapper into AgentScope

### PR DESCRIPTION
**Pull request by qbc2016**
-) Raise RuntimeError when the response of Tongyi is not valid
-) preprocess the roles in messages before calling Tongyi model

### CODE REVIEW

**Review by DavdGao — COMMENTED** [review]
Please see inline comments

File: src/agentscope/models/tongyi_model.py [hunk 1]
```diff
@@ -200,18 +202,63 @@ def __call__(
...
-            self.monitor.update(
-                response.usage,
-                prefix=self.model,
-            )
-        except QuotaExceededError as e:
-            # TODO: optimize quota exceeded error handling process
-            logger.error(e.message)
+        # try:
+        #     self.monitor.update(
+        #         response.usage,
```
[the remark this answers is not in the document]
Comment (qbc2016) [reply src/agentscope/models/tongyi_model.py:209]: yes, e.g., it contains the following:  "usage": {"input_tokens": 25, "output_tokens": 31, "total_tokens": 56}

File: src/agentscope/models/tongyi_model.py [hunk 2]
```diff
@@ -200,18 +202,63 @@ def __call__(
...
+            error_msg = (
+                f"Request id: {response.request_id},"
+                f" Status code: {response.status_code},"
+                f" error code: {response.code},"
+                f" error message: {response.message}."
+            )
+
+            raise RuntimeError(error_msg)
+
+    def _preprocess_role(self, messages: list) -> list:
```
[the remark this answers is not in the document]
Comment (qbc2016) [reply src/agentscope/models/tongyi_model.py:231]: Do you mean that we add a new PromptType, such as PromptType.Qwen? Then check if the instance of the model is QwenWrapper, if it is, then do the preprocess_role in the join function of PromptEngine?

File: src/agentscope/models/tongyi_model.py [hunk 3]
```diff
@@ -13,9 +14,13 @@
...
-from ..utils import QuotaExceededError
 from ..constants import _DEFAULT_API_BUDGET
 
+# The models in this list require that the roles of messages must alternate
+# between "user" and "assistant".
+# TODO: add more models
+SPECIAL_MODEL_LIST = ["qwen-turbo", "qwen-plus", "qwen1.5-72b-chat"]
+
 
 class TongyiWrapper(ModelWrapperBase):
```
[the remark this answers is not in the document]
Comment (qbc2016) [reply src/agentscope/models/tongyi_model.py:25]: There may be Tongyiwanxiang and more models based on this TongyiWrapper, I prefer to leave it and change TongyiChatWrapper to QwenChatWrapper

[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/models/tongyi_model.py:25]: The naming is a little strange. 
1. First, `dashscope.Generation.call` supports more than qwen models, so `QwenChatWrapper` is inappropriate. Maybe `DashScopeChatWrapper`, and the base model `TongyiWrapper` should be `DashScopeWrapperBase`? (It's for DashScope, not qwen)
2. As stated above, next we can support more APIs in DashScope like `dashscope.TextEmbedding.call` by `DashScopeTextEmbeddingWrapper` in AgentScope.

File: src/agentscope/models/tongyi_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/models/tongyi_model.py:223]: So we can add a new metric to record the token usage in monitor here.

**Review by DavdGao — COMMENTED** [review]
Plz see inline comments

File: src/agentscope/models/tongyi_model.py [hunk 4]
```diff
@@ -13,9 +14,13 @@
 
 from ..utils.monitor import MonitorFactory
 from ..utils.monitor import get_full_name
-from ..utils import QuotaExceededError
 from ..constants import _DEFAULT_API_BUDGET
 
+# The models in this list require that the roles of messages must alternate
+# between "user" and "assistant".
+# TODO: add more models
+SPECIAL_MODEL_LIST = ["qwen-turbo", "qwen-plus", "qwen1.5-72b-chat"]
```
Comment (pan-x-c) [inline src/agentscope/models/tongyi_model.py:22]: qwen-max seems to have the same problem.
https://help.aliyun.com/zh/dashscope/developer-reference/api-details?spm=a2c4g.11186623.0.i22

Comment (qbc2016) [reply src/agentscope/models/tongyi_model.py:22]: yes, only qwen-max-1201 needs that the first and the last role should be users

**Review by DavdGao — COMMENTED** [review]
Plz see inline comments

File: src/agentscope/models/dashscope_model.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    """The model wrapper for DashScope's chat API."""
+
+    model_type: str = "dashscope_chat"
+
+    def _register_default_metrics(self) -> None:
+        # Set monitor accordingly
+        # TODO: set quota to the following metrics
+        self.monitor = MonitorFactory.get_monitor()
+        self.monitor.register(
+            self._metric("input_tokens"),
```
[the remark this answers is not in the document]
Comment (qbc2016) [reply src/agentscope/models/dashscope_model.py:114]: for openai, the usage term is "usage": {"prompt_tokens": x, "completion_tokens": x, "total_tokens": x}, but for qwen chat, the usage term is "usage": {"input_tokens": x, "output_tokens": x, "total_tokens": x},

File: src/agentscope/models/dashscope_model.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+                "messages": messages,
+                **kwargs,
+            },
+            json_response=response,
+        )
+
+        # step5: update monitor accordingly
+        try:
+            self.monitor.update(
+                response.usage,
```
[the remark this answers is not in the document]
Comment (qbc2016) [reply src/agentscope/models/dashscope_model.py:215]: See above, I think they are matched

File: src/agentscope/models/dashscope_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/models/dashscope_model.py:114]: Yes, but if we want to compare the usage of different models in monitor, we must align their metric names. e.g. we may want to print the token usage of all different models (qwen, gpt-4, etc)

[the remark this answers is not in the document]
Comment (qbc2016) [reply src/agentscope/models/dashscope_model.py:114]: ok, I will modify it in the similar way you did in Gemini chat

File: src/agentscope/models/dashscope_model.py [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+        if save_local:
+            # Return local url if save_local is True
+            urls = [file_manager.save_image(_) for _ in urls]
+        return ModelResponse(image_urls=urls, raw=response)
+
+
+class DashScopeTextEmbeddingWrapper(DashScopeWrapper):
+    """The model wrapper for DashScope Text Embedding API."""
+
+    model_type: str = "dashscope_textembedding"
```
Comment (DavdGao) [inline src/agentscope/models/dashscope_model.py:336]: maybe unified into `dashscope_embedding`

File: src/agentscope/models/dashscope_model.py [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+        return ModelResponse(
+            text=response.output["choices"][0]["message"]["content"],
+            raw=response,
+        )
+
+
+class DashScopeImageSynthesisWrapper(DashScopeWrapper):
+    """The model wrapper for DashScope Image Synthesis API."""
+
+    model_type: str = "dashscope_imagesynthesis"
```
Comment (DavdGao) [inline src/agentscope/models/dashscope_model.py:235]: maybe "dashscope_image_synthesis" or "dashscope_image"?

**Review by DavdGao — COMMENTED** [review]
please see inline comments, others lgtm

File: src/agentscope/models/dashscope_model.py (hunk 7 above)
Comment (qbc2016) [reply src/agentscope/models/dashscope_model.py]: since dashscope has TextEmbedding and MultiModalEmbedding, I change it to dashscope_text_embedding

**Review by pan-x-c — APPROVED** [review]
LGTM

**Pull request merged**